cardinal/server/tx.go

```diff
@@ -1,127 +1,144 @@
+	"fmt"
+	"github.com/gofiber/fiber/v2"
 	"pkg.world.dev/world-engine/cardinal/ecs"
-
+	"pkg.world.dev/world-engine/cardinal/types/message"
 	"pkg.world.dev/world-engine/sign"
 )
 
-func (handler *Handler) processTransaction(tx message.Message, payload []byte, sp *sign.Transaction,
-) (*TransactionReply, error) {
-	txVal, err := tx.Decode(payload)
-	if err != nil {
-		return nil, eris.Wrap(err, "unable to decode transaction")
-	}
-	return handler.submitTransaction(txVal, tx, sp)
-}
+var (
+	ErrNoPersonaTag               = errors.New("persona tag is required")
+	ErrWrongNamespace             = errors.New("incorrect namespace")
+	ErrSystemTransactionRequired  = errors.New("system transaction required")
+	ErrSystemTransactionForbidden = errors.New("system transaction forbidden")
+)
 
-func getTxFromParams(pathParam string, params interface{}, txNameToTx map[string]message.Message,
-) (message.Message, error) {
-	mappedParams, ok := params.(map[string]interface{})
-	if !ok {
-		return nil, eris.New("params not readable")
-	}
-	txType, ok := mappedParams[pathParam]
-	if !ok {
-		return nil, eris.New("params do not contain txType from the path /tx/game/{txType}")
-	}
-	txTypeString, ok := txType.(string)
-	if !ok {
-		return nil, eris.New("txType needs to be a string from path")
-	}
-	tx, ok := txNameToTx[txTypeString]
-	if !ok {
-		return nil, eris.Errorf("could not locate transaction type: %s", txTypeString)
-	}
-	return tx, nil
+// TransactionReply is the type that is sent back to clients after a transaction is added to the queue.
+type TransactionReply struct {
+	TxHash string
+	Tick   uint64
 }
 
-func (handler *Handler) getBodyAndSigFromParams(
-	params interface{},
-	isSystemTransaction bool) ([]byte, *sign.Transaction, error) {
-	mappedParams, ok := params.(map[string]interface{})
-	if !ok {
-		return nil, nil, eris.New("params not readable")
-	}
-	txBody, ok := mappedParams["txBody"]
-	if !ok {
-		return nil, nil, eris.New("params do not contain txBody from the body of the http request")
-	}
-	txBodyMap, ok := txBody.(map[string]interface{})
-	if !ok {
-		return nil, nil, eris.New("txBody needs to be a json object in the body")
-	}
-	payload, sp, err := handler.verifySignatureOfMapRequest(txBodyMap, isSystemTransaction)
+func (s *Server) registerTransactionHandler(path string) error {
+	msgs, err := s.eng.ListMessages()
 	if err != nil {
-		return nil, nil, eris.Wrap(err, "error verifying signature of map request")
+		return err
 	}
-	return payload, sp, nil
-}
 
-// register transaction handlers on swagger server.
-func (handler *Handler) registerTxHandlerSwagger(api *untyped.API) error {
-	world := handler.w
-	txs, err := world.ListMessages()
-	if err != nil {
-		return err
+	// some messages may have a custom path. we store them separately.
+	msgNameToMsg := make(map[string]message.Message)
+	customPathToMsg := make(map[string]message.Message)
+	for _, msg := range msgs {
+		if msg.Path() == "" {
+			msgNameToMsg[msg.Name()] = msg
```

Comment:
nit, and may involve code changes in places where you didn't touch. 

Better to store a pointer to avoid triggering copy overhead.

Reply:
these are interfaces, so they are pointers